File: content-solutions/binary-search/tree-saw.js
function solution(trees, m) {
  let lo = 0;
  let hi = Math.max(...trees);
  let answer = 0;
  while (lo <= hi) {
    const mid = Math.floor((lo + hi) / 2);
    let got = 0;
    for (const t of trees) if (t > mid) got += t - mid;
    if (got >= m) {
      answer = mid;
      lo = mid + 1;
    } else hi = mid - 1;
  }
  return answer;
}

File: content-solutions/binary-search/tree-saw.java
class Solution {
    public int solution(int[] trees, int m) {
        long lo = 0, hi = 0, answer = 0;
        for (int t : trees) hi = Math.max(hi, t);
        while (lo <= hi) {
            long mid = (lo + hi) / 2;
            long got = 0;
            for (int t : trees) if (t > mid) got += t - mid;
            if (got >= m) {
                answer = mid;
                lo = mid + 1;
            } else hi = mid - 1;
        }
        return (int) answer;
    }
}
